libres/lib/python/logging.cpp

```diff
@@ -0,0 +1,134 @@
+#include <pybind11/stl.h>
+
+#include <ert/python.hpp>
+
+using namespace std::string_literals;
+namespace py = pybind11;
+
+// Controls the prefix of the Python logger:
+//   logging.getLogger(f"{LOGGER_NAMESPACE}.{name}")
+#define LOGGER_NAMESPACE "res"s
+
+namespace {
+class Logger : public ert::ILogger {
+    struct interface {
+        py::object debug;
+        py::object info;
+        py::object warning;
+        py::object error;
+        py::object critical;
+
+        interface(py::object logger);
+    };
+
+    std::unique_ptr<interface> m_interface;
+
+public:
+    void init(const std::string &name);
+
+    /** Called on Python shutdown to release the py::objects */
+    void shutdown() { m_interface = nullptr; }
+
+protected:
+    void log(Level level, fmt::string_view f, fmt::format_args args) final;
+};
+
+bool has_init_logging{};
+std::unordered_map<std::string, std::shared_ptr<Logger>> loggers;
+} // namespace
+
+/**
+ * Fetch functions from Python. Note that we keep hold of the functions, but
+ * they can still get monkeypatched by eg. pytest's caplog.
+ */
+Logger::interface::interface(py::object logger)
+    : debug(logger.attr("debug")), info(logger.attr("info")),
+      warning(logger.attr("warning")), error(logger.attr("error")),
+      critical(logger.attr("critical")) {}
+
+void Logger::init(const std::string &name) {
+    auto get_logger = py::module_::import("logging").attr("getLogger");
+
+    // Don't include the . for the default logger (whose name is "")
+    auto full_name =
+        name.empty() ? LOGGER_NAMESPACE : LOGGER_NAMESPACE "."s + name;
+    auto logger = get_logger(full_name);
+    m_interface = std::make_unique<interface>(logger);
+}
+
+void Logger::log(Logger::Level level, fmt::string_view f,
+                 fmt::format_args args) {
+    if (!m_interface) {
+        // Python has either not initalised us yet or has shut down. Drop logs.
+        return;
+    }
+
+    // Calling via cwrap does not acquire GIL, so do it now
+    py::gil_scoped_acquire gil;
+
+    auto payload = fmt::vformat(f, args);
+    switch (level) {
+    case Level::debug:
+        m_interface->debug(payload);
+        break;
+    case Level::info:
+        m_interface->info(payload);
+        break;
+    case Level::warning:
+        m_interface->warning(payload);
+        break;
+    case Level::error:
+        m_interface->error(payload);
+        break;
+    case Level::critical:
+        m_interface->critical(payload);
+        break;
+    }
+}
+
+std::shared_ptr<ert::ILogger> ert::get_logger(const std::string &name) {
+    auto it = loggers.find(name);
+    if (it != loggers.end())
+        return it->second;
+
+    auto logger = std::make_shared<Logger>();
+    loggers[name] = logger;
+
+    if (has_init_logging)
+        logger->init(name);
+
+    return logger;
+}
+
+void init_logging(py::module_ m) {
+    has_init_logging = true;
+
+    // Initialise all loggers that were created before Python initialised
```

Comment:
Only necessary when loggers can be defined at compile time - i.e. if the documentation in https://github.com/equinor/ert/pull/2550#issuecomment-991268694 is still valid with this new impl.

If not, `has_init_logging` as well as the `unordered_map loggers` can go away since we will (most likely!) never create logger-objects at runtime prior to init from Python.

Reply:
If we go by the principle of least surprise, then I think this is necessary. If someone tries to have a globally defined logger as is the norm in Python, the program would crash or worse, be in a broken state and then crash sometime don't the line for no obvious reason.

Not sure what you mean by removing the `unordered_map loggers`. We need that if you want more than 1 logger.

Reply:
Mmm...  didn't read the logic in `get_logger` properly...  you essentially maintain a set of global objects in C++ with each linked to a corresponding Python logger-object, linked in the `init()` -call.

I don't completely see limitations and/or implications, but wouldn't it be simpler to not link C++ objects with corresponding Python-objects (eliminating the need for maintaining this set in C++) and just use the fwd-approach from https://github.com/equinor/ert/pull/2500 (looking up the right Python-logger in each log-call)?

As mentioned: I don't completely see limitations of the simpler approach vs your approach so feel free to point out any. 😄  One point I do see is performance, but I would claim that this is not really a concern although I'm open to discuss it.

Reply:
It's a performance consideration, yes. I'm not particularly happy with C++ that would be slower than equivalent Python. In Python you have one logger instance per module which you wouldn't lookup every time you call `logger.debug`.

Reply:
I'm willing to bet quite a lot that the performance-hit of calling `getLogger` second time for some name is pretty small since I assume they use some kind of hashmap. The [logging-docs on optimisation](https://docs.python.org/2/howto/logging.html#optimization) hints at very different sources for slow log-calls.

Don't get me wrong - I don't think it is elegant to retrieve the logger-object in every call, and I would certainly not do that when writing pure Python-code. The idea here is that doing this "a llittle stupid" can make the glue-code a lot simpler with only a reasonable performance-hit.

Having said this, I do think your solution is nice and look forward to have this functionality available!

Reply:
Ref our discussion on (theoretical) speed vs less shared data: In this case I would personally sacrifice the former for the latter, but I respect your opposite view and will approve this PR to get functionality merged.